stats/statsd/shared/collector.go

```diff
@@ -0,0 +1,200 @@
+/*
+ *
+ * k6 - a next-generation load testing tool
+ * Copyright (C) 2016 Load Impact
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package statsd
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/DataDog/datadog-go/statsd"
+	"github.com/loadimpact/k6/lib"
+	"github.com/loadimpact/k6/stats"
+	"github.com/loadimpact/k6/ui"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	pushInterval = 1 * time.Second
+)
+
+var _ lib.Collector = &Collector{}
+
+// Tagger defines a type that can process tags
+type Tagger interface {
+	Process(string) func(map[string]string, string) []string
+}
+
+// Collector defines a collector struct
+type Collector struct {
+	Client  *statsd.Client
+	Config  Config
+	Logger  *log.Entry
+	Type    ClientType
+	Tagger  Tagger
+	Summary map[string]*stats.Metric
+
+	startTime  time.Time
+	buffer     []*Sample
+	bufferLock sync.Mutex
+}
+
+// Init sets up the collector
+func (c *Collector) Init() error {
+	c.Summary = make(map[string]*stats.Metric)
+	return nil
+}
+
+// Link returns the address of the client
+func (c *Collector) Link() string {
+	return c.Config.Address()
+}
+
+// Run the collector
+func (c *Collector) Run(ctx context.Context) {
+	c.Logger.Debugf("%s: Running!", c.Type.String())
```

Comment:
Use `WithField()`/`WithFields()` rather than `Levelf()`.

Reply:
Not sure what you mean here. Type is used to prefix the string not as a meta/tagging. I guess if that goes again't any project convention I'll have to change all of the other similar instances.